File: sdk/android/mosaic/src/main/kotlin/dev/mosaic/sdk/TransactionObservation.kt
package dev.mosaic.sdk

import com.google.gson.JsonObject
import com.google.gson.JsonParser

/**
 * Optional, opt-in Transaction Observation handoff (Billing Ingestion Contract 1).
 *
 * A Transaction Observation is a *trigger*, never proof. The SDK reports a bounded provider
 * reference so Mosaic's server can start its own validation with its own credentials sooner than a
 * store notification would arrive. Nothing in this file can claim, imply, or be read as a validated
 * transaction: there is deliberately no result member named validated, verified or confirmed, and a
 * submission outcome never changes a local [MosaicPurchaseResult].
 *
 * The frozen contract under `protocol/schema/billing-ingestion/v1/` is the authority for both the
 * request and the response record; the canonical fixture
 * `protocol/fixtures/billing-ingestion/v1/google-client-observation.json` is asserted verbatim.
 *
 * What is structurally impossible to carry here: a Google Play purchase token, `originalJson`, a
 * purchase signature, obfuscated account or profile identifiers, any credential, and any
 * client-asserted Store Environment — classification is a server-side decision made from verified
 * provider metadata, and the contract rejects a client observation that asserts one.
 */
sealed interface MosaicTransactionObservationResult {
    /** The observation is well formed and queued for server-side validation. Nothing more. */
    data object AcceptedForValidation : MosaicTransactionObservationResult

    /** The identical submission was already recorded. Idempotent, not an error. */
    data object Duplicate : MosaicTransactionObservationResult

    /** Resubmitting the identical document cannot succeed. */
    data class PermanentlyRejected(val code: String) : MosaicTransactionObservationResult

    /** Transient. The identical document is resubmitted later from the durable queue. */
    data class RetryableFailure(
        val code: String,
        val retryAfterSeconds: Int? = null,
    ) : MosaicTransactionObservationResult
}

internal const val MOSAIC_BILLING_INGESTION_CONTRACT_VERSION = "1"
internal const val MOSAIC_RECORD_CLIENT_OBSERVATION = "clientTransactionObservation"
internal const val MOSAIC_RECORD_SUBMISSION_RESULT = "observationSubmissionResult"

/** Reference kinds for the Google Play token digest and order handle, frozen by the contract. */
internal const val MOSAIC_REFERENCE_GOOGLE_PLAY_TOKEN_DIGEST = "google_play_token_digest"
internal const val MOSAIC_REFERENCE_GOOGLE_PLAY_ORDER_ID = "google_play_order_id"
internal const val MOSAIC_STORE_PLATFORM_GOOGLE_PLAY = "google_play"

/**
 * The only authority a client may ever assert. It is emitted by the codec rather than carried as a
 * settable field, so no call site can raise this SDK's authority above an untrusted observation.
 */
internal const val MOSAIC_SOURCE_AUTHORITY_CLIENT = "client_observation"

/** SDK context, mirroring the Analytics Event context vocabulary. It carries no tenant identity. */
internal data class MosaicTransactionObservationContext(
    val platform: String = "android",
    val sdkFamily: String = "android",
    val sdkVersion: String = MOSAIC_ANDROID_SDK_VERSION,
    val applicationVersion: String? = null,
    val operatingSystemVersion: String? = null,
)

/** Correlation to Analytics Events uses the existing opaque handles only. */
internal data class MosaicTransactionObservationCorrelation(
    val purchaseAttemptId: String? = null,
    val providerUpdateId: String? = null,
) {
    val isEmpty: Boolean get() = purchaseAttemptId == null && providerUpdateId == null
}

/**
 * One submittable observation. This is the complete set of values that ever leaves the device for
 * this feature; there is no free-form field, no provider payload, and no host-supplied text.
 */
internal data class MosaicTransactionObservation(
    /** Stable record identity, generated once and persisted; unchanged across every retry. */
    val observationId: String,
    /** Deterministic deduplication key, derived from the reference and the outcome. */
    val submissionId: String,
    val providerId: String,
    val referenceKind: String,
    val reference: String,
    val providerOrderReference: String? = null,
    val observedAt: String,
    val context: MosaicTransactionObservationContext,
    val correlation: MosaicTransactionObservationCorrelation? = null,
    /** A claim only. The server resolves the Mosaic Product independently. */
    val claimedMosaicProductId: String? = null,
    val storePlatform: String = MOSAIC_STORE_PLATFORM_GOOGLE_PLAY,
)

/** Bounds mirroring the frozen contract. Everything else is rejected before it can be queued. */
internal object MosaicTransactionObservationBounds {
    private val IDENTIFIER = Regex("^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$")
    private val GOOGLE_TOKEN_DIGEST = Regex("^[a-f0-9]{64}$")
    private val TIMESTAMP =
        Regex("^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(?:\\.[0-9]{1,6})?Z$")
    private val VERSION = Regex("^[A-Za-z0-9][A-Za-z0-9.+_-]{0,63}$")
    private val OPERATING_SYSTEM_VERSION = Regex("^[A-Za-z0-9][A-Za-z0-9.+_ -]{0,63}$")

    /** `providerOrderReference` value pattern; a purchase token cannot satisfy the 128 bound. */
    private val ORDER_REFERENCE = Regex("^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$")

    fun isValidIdentifier(value: String): Boolean = IDENTIFIER.matches(value)

    fun isValidReference(referenceKind: String, value: String): Boolean = when (referenceKind) {
        MOSAIC_REFERENCE_GOOGLE_PLAY_TOKEN_DIGEST -> GOOGLE_TOKEN_DIGEST.matches(value)
        else -> false
    }

    fun isValidOrderReference(value: String): Boolean = ORDER_REFERENCE.matches(value)

    fun isValidTimestamp(value: String): Boolean = TIMESTAMP.matches(value)

    fun isSubmittable(observation: MosaicTransactionObservation): Boolean =
        isValidIdentifier(observation.observationId) &&
            isValidIdentifier(observation.submissionId) &&
            isValidIdentifier(observation.providerId) &&
            observation.storePlatform == MOSAIC_STORE_PLATFORM_GOOGLE_PLAY &&
            isValidReference(observation.referenceKind, observation.reference) &&
            isValidTimestamp(observation.observedAt) &&
            (observation.providerOrderReference?.let(::isValidOrderReference) != false) &&
            (observation.claimedMosaicProductId?.let(::isValidIdentifier) != false) &&
            isValidContext(observation.context) &&
            isValidCorrelation(observation.correlation)

    private fun isValidContext(context: MosaicTransactionObservationContext): Boolean =
        context.platform == "android" &&
            context.sdkFamily == "android" &&
            VERSION.matches(context.sdkVersion) &&
            (context.applicationVersion?.let(VERSION::matches) != false) &&
            (context.operatingSystemVersion?.let(OPERATING_SYSTEM_VERSION::matches) != false)

    private fun isValidCorrelation(correlation: MosaicTransactionObservationCorrelation?): Boolean =
        correlation == null ||
            (
                !correlation.isEmpty &&
                    (correlation.purchaseAttemptId?.let(::isValidIdentifier) != false) &&
                    (correlation.providerUpdateId?.let(::isValidIdentifier) != false)
                )
}

/**
 * Field-by-field codec for the contract's record envelopes. Nothing is bound reflectively, so no
 * consumer R8 keep rule is required and no field can reach the wire by accident.
 */
internal object MosaicTransactionObservationCodec {
    private val STATUSES = setOf(
        "accepted_for_validation",
        "duplicate",
        "permanently_rejected",
        "retryable_failure",
    )

    fun encode(observation: MosaicTransactionObservation): String = JsonObject().apply {
        addProperty("billingIngestionContractVersion", MOSAIC_BILLING_INGESTION_CONTRACT_VERSION)
        addProperty("recordType", MOSAIC_RECORD_CLIENT_OBSERVATION)
        add("payload", encodePayload(observation))
    }.toString()

    private fun encodePayload(observation: MosaicTransactionObservation): JsonObject = JsonObject().apply {
        addProperty("observationId", observation.observationId)
        addProperty("submissionId", observation.submissionId)
        addProperty("providerId", observation.providerId)
        addProperty("storePlatform", observation.storePlatform)
        add("transactionReference", reference(observation.referenceKind, observation.reference))
        observation.providerOrderReference?.let {
            add("providerOrderReference", reference(MOSAIC_REFERENCE_GOOGLE_PLAY_ORDER_ID, it))
        }
        addProperty("observedAt", observation.observedAt)
        addProperty("sourceAuthority", MOSAIC_SOURCE_AUTHORITY_CLIENT)
        add(
            "context",
            JsonObject().apply {
                addProperty("platform", observation.context.platform)
                addProperty("sdkFamily", observation.context.sdkFamily)
                addProperty("sdkVersion", observation.context.sdkVersion)
                observation.context.applicationVersion?.let { addProperty("applicationVersion", it) }
                observation.context.operatingSystemVersion?.let { addProperty("operatingSystemVersion", it) }
            },
        )
        // The contract requires at least one member, so an empty correlation is omitted entirely.
        observation.correlation?.takeIf { !it.isEmpty }?.let { correlation ->
            add(
                "correlation",
                JsonObject().apply {
                    correlation.purchaseAttemptId?.let { addProperty("purchaseAttemptId", it) }
                    correlation.providerUpdateId?.let { addProperty("providerUpdateId", it) }
                },
            )
        }
        observation.claimedMosaicProductId?.let { addProperty("claimedMosaicProductId", it) }
    }

    private fun reference(kind: String, value: String): JsonObject = JsonObject().apply {
        addProperty("referenceKind", kind)
        addProperty("value", value)
    }

    fun decode(source: String): MosaicTransactionObservation {
        val root = JsonParser.parseString(source).asJsonObject
        require(root.keySet() == setOf("billingIngestionContractVersion", "recordType", "payload"))
        require(
            root.get("billingIngestionContractVersion").asString ==
                MOSAIC_BILLING_INGESTION_CONTRACT_VERSION,
        )
        require(root.get("recordType").asString == MOSAIC_RECORD_CLIENT_OBSERVATION)
        val payload = root.getAsJsonObject("payload")
        require(
            payload.keySet().all {
                it in setOf(
                    "observationId", "submissionId", "providerId", "storePlatform",
                    "transactionReference", "providerOrderReference", "observedAt",
                    "sourceAuthority", "context", "correlation", "claimedMosaicProductId",
                )
            },
        )
        require(payload.get("sourceAuthority").asString == MOSAIC_SOURCE_AUTHORITY_CLIENT)
        val transactionReference = payload.getAsJsonObject("transactionReference")
        val orderReference = payload.getAsJsonObject("providerOrderReference")
        require(
            orderReference == null ||
                orderReference.get("referenceKind").asString == MOSAIC_REFERENCE_GOOGLE_PLAY_ORDER_ID,
        )
        val context = payload.getAsJsonObject("context")
        require(
            context.keySet().all {
                it in setOf(
                    "platform", "sdkFamily", "sdkVersion", "applicationVersion", "operatingSystemVersion",
                )
            },
        )
        val correlation = payload.getAsJsonObject("correlation")
        require(
            correlation == null ||
                correlation.keySet().all { it in setOf("purchaseAttemptId", "providerUpdateId") },
        )
        val observation = MosaicTransactionObservation(
            observationId = payload.get("observationId").asString,
            submissionId = payload.get("submissionId").asString,
            providerId = payload.get("providerId").asString,
            storePlatform = payload.get("storePlatform").asString,
            referenceKind = transactionReference.get("referenceKind").asString,
            reference = transactionReference.get("value").asString,
            providerOrderReference = orderReference?.get("value")?.asString,
            observedAt = payload.get("observedAt").asString,
            context = MosaicTransactionObservationContext(
                platform = context.get("platform").asString,
                sdkFamily = context.get("sdkFamily").asString,
                sdkVersion = context.get("sdkVersion").asString,
                applicationVersion = context.get("applicationVersion")?.asString,
                operatingSystemVersion = context.get("operatingSystemVersion")?.asString,
            ),
            correlation = correlation?.let {
                MosaicTransactionObservationCorrelation(
                    purchaseAttemptId = it.get("purchaseAttemptId")?.asString,
                    providerUpdateId = it.get("providerUpdateId")?.asString,
                )
            },
            claimedMosaicProductId = payload.get("claimedMosaicProductId")?.asString,
        )
        require(MosaicTransactionObservationBounds.isSubmittable(observation))
        return observation
    }

    /**
     * Reads the contract's submission-result record. An unrecognized record type, contract version,
     * or status is not decoded: an observation is only ever removed from the queue on a result this
     * SDK understands, so a future, malformed, or spoofed response retries rather than silently
     * discarding the handoff. The status set has no member meaning validated.
     */
    fun decodeResult(source: String, submissionId: String): MosaicTransactionObservationResult? {
        val payload = runCatching {
            val root = JsonParser.parseString(source).asJsonObject
            if (root.get("billingIngestionContractVersion")?.asString !=
                MOSAIC_BILLING_INGESTION_CONTRACT_VERSION
            ) {
                return null
            }
            if (root.get("recordType")?.asString != MOSAIC_RECORD_SUBMISSION_RESULT) return null
            root.getAsJsonObject("payload")
        }.getOrNull() ?: return null
        val echoed = runCatching { payload.get("submissionId")?.asString }.getOrNull()
        if (echoed != null && echoed != submissionId) return null
        val status = runCatching { payload.get("status")?.asString }.getOrNull()
            ?.takeIf { it in STATUSES } ?: return null
        val code = runCatching { payload.get("code")?.asString }.getOrNull()
            ?.takeIf { MosaicTransactionObservationBounds.isValidIdentifier(it) }
        // The body's retryAfterSeconds is the authority; a transport header is only the fallback.
        val retryAfterSeconds = runCatching { payload.get("retryAfterSeconds")?.asInt }.getOrNull()
            ?.takeIf { it in 1..86_400 }
        return when (status) {
            "accepted_for_validation" -> MosaicTransactionObservationResult.AcceptedForValidation
            "duplicate" -> MosaicTransactionObservationResult.Duplicate
            "permanently_rejected" ->
                MosaicTransactionObservationResult.PermanentlyRejected(code ?: "observation_schema_invalid")
            else -> MosaicTransactionObservationResult.RetryableFailure(
                code ?: "service_temporarily_unavailable",
                retryAfterSeconds,
            )
        }
    }
}

File: sdk/ios/Sources/MosaicSDK/TransactionObservation.swift
import Foundation

/// Whether the SDK hands observed provider transactions to Mosaic for
/// server-side validation.
///
/// A Transaction Observation is a trigger, never proof. Enabling this never
/// changes a local purchase result, never grants access, and never re-labels a
/// StoreKit-verified purchase as server-validated.
public enum MosaicTransactionObservationMode: Sendable, Equatable {
  /// The default. No observation is built, queued, persisted, or sent.
  case disabled
  case enabled
}

/// How a submitted `reference` must be read by the server.
///
/// iOS submits only the raw decimal App Store transaction identifier. The
/// remaining Billing Ingestion Contract 1 reference kinds belong to other
/// store platforms and are deliberately absent here.
public enum MosaicTransactionObservationReferenceKind: String, Sendable, Equatable, Codable {
  case appStoreTransactionID = "app_store_transaction_id"
}

/// Which store a reference belongs to. A record's store platform must match
/// its reference kind, so it is derived rather than supplied.
public enum MosaicTransactionObservationStorePlatform: String, Sendable, Equatable, Codable {
  case appleAppStore = "apple_app_store"
}

/// Correlation to Analytics Event 1/2 through the existing opaque handles
/// only. No new cross-contract identifier is introduced.
public struct MosaicTransactionObservationCorrelation: Sendable, Equatable {
  public let purchaseAttemptID: String?
  public let providerOperationID: String?
  public let providerUpdateID: String?

  /// Returns `nil` when no handle is present or any handle is not a contract
  /// identifier, because the contract requires at least one valid member.
  public init?(
    purchaseAttemptID: String? = nil,
    providerOperationID: String? = nil,
    providerUpdateID: String? = nil
  ) {
    let handles = [purchaseAttemptID, providerOperationID, providerUpdateID].compactMap { $0 }
    guard !handles.isEmpty,
      handles.allSatisfy(MosaicTransactionObservation.isIdentifier)
    else { return nil }
    self.purchaseAttemptID = purchaseAttemptID
    self.providerOperationID = providerOperationID
    self.providerUpdateID = providerUpdateID
  }

  var wireValue: [String: String] {
    var value: [String: String] = [:]
    if let purchaseAttemptID { value["purchaseAttemptId"] = purchaseAttemptID }
    if let providerOperationID { value["providerOperationId"] = providerOperationID }
    if let providerUpdateID { value["providerUpdateId"] = providerUpdateID }
    return value
  }
}

extension MosaicTransactionObservationCorrelation: Codable {
  enum CodingKeys: String, CodingKey {
    case purchaseAttemptID = "purchaseAttemptId"
    case providerOperationID = "providerOperationId"
    case providerUpdateID = "providerUpdateId"
  }
}

/// One untrusted client report that a provider transaction may exist.
///
/// The value type carries no receipt, no signed payload, no device
/// verification material, no account token, no tenant identity, and no Store
/// Environment. Tenant scope is derived server-side from the authenticated
/// public SDK key, and Store Environment is classified server-side during
/// validation: a client never asserts one.
public struct MosaicTransactionObservation: Sendable, Equatable, Codable {
  /// Identity of this observation record. It is derived from the submission
  /// identifier so a replay after relaunch reproduces the identical record
  /// instead of colliding with itself.
  public let observationID: String
  /// Deterministic idempotency key. It is never derived from a timestamp,
  /// price, Product, or subject, so a retry after relaunch deduplicates.
  public let submissionID: String
  /// The Mosaic commerce provider identity, matching the installed adapter.
  public let providerID: String
  public let referenceKind: MosaicTransactionObservationReferenceKind
  /// The raw provider reference, carried verbatim as a string. It is never
  /// parsed into an integer or a floating-point number: App Store transaction
  /// identifiers exceed both IEEE-754 double precision and `Int64`.
  public let reference: String
  public let observedAt: Date
  public let correlation: MosaicTransactionObservationCorrelation?

  /// Derived from the reference kind, never supplied, so a record can never
  /// claim a platform its reference does not belong to.
  public var storePlatform: MosaicTransactionObservationStorePlatform {
    switch referenceKind {
    case .appStoreTransactionID: .appleAppStore
    }
  }

  /// Returns `nil` when the reference or an identifier would violate the
  /// Billing Ingestion Contract 1 bounds, so a malformed value is dropped on
  /// the device rather than rejected at ingest.
  public init?(
    submissionID: String,
    referenceKind: MosaicTransactionObservationReferenceKind = .appStoreTransactionID,
    reference: String,
    providerID: String = "app_store",
    observationID: String? = nil,
    observedAt: Date = Date(),
    correlation: MosaicTransactionObservationCorrelation? = nil
  ) {
    let identity = observationID ?? "observation_\(submissionID)"
    guard Self.isIdentifier(submissionID), Self.isIdentifier(identity),
      Self.isIdentifier(providerID), Self.isValidReference(reference, kind: referenceKind)
    else { return nil }
    self.observationID = identity
    self.submissionID = submissionID
    self.providerID = providerID
    self.referenceKind = referenceKind
    self.reference = reference
    self.correlation = correlation
    // Normalizing through the wire representation keeps a persisted record
    // equal to the record it was restored from.
    self.observedAt =
      MosaicAnalyticsRuntime.parseTimestamp(MosaicAnalyticsRuntime.timestamp(observedAt))
      ?? observedAt
  }

  static func isIdentifier(_ value: String) -> Bool {
    value.count <= 128
      && value.range(of: "^[A-Za-z0-9][A-Za-z0-9._:-]*$", options: .regularExpression) != nil
  }

  static func isValidReference(
    _ value: String, kind: MosaicTransactionObservationReferenceKind
  ) -> Bool {
    switch kind {
    case .appStoreTransactionID:
      value.range(of: "^[0-9]{1,24}$", options: .regularExpression) != nil
    }
  }
}

/// SDK context, mirroring the Analytics Event context vocabulary. It carries
/// no Organization, Project, Environment, or Application identity.
struct MosaicTransactionObservationContext: Sendable, Equatable, Codable {
  var platform: String { "ios" }
  var sdkFamily: String { "ios" }
  let sdkVersion: String
  let operatingSystemVersion: String?
  let applicationVersion: String?

  init(sdkVersion: String = mosaicSDKVersion, applicationVersion: String?) {
    self.sdkVersion = Self.bounded(sdkVersion) ?? "0"
    operatingSystemVersion = Self.bounded(
      ProcessInfo.processInfo.operatingSystemVersionString
        .split(separator: " ").first(where: { $0.first?.isNumber == true }).map(String.init))
    self.applicationVersion = Self.bounded(applicationVersion)
  }

  /// The contract bounds these to 64 printable characters with no leading
  /// punctuation. A host-supplied value that does not qualify is omitted
  /// rather than sent and rejected.
  private static func bounded(_ value: String?) -> String? {
    guard let value, value.count <= 64,
      value.range(of: "^[A-Za-z0-9][A-Za-z0-9.+_-]*$", options: .regularExpression) != nil
    else { return nil }
    return value
  }

  var wireValue: [String: String] {
    var value = ["platform": platform, "sdkFamily": sdkFamily, "sdkVersion": sdkVersion]
    if let operatingSystemVersion { value["operatingSystemVersion"] = operatingSystemVersion }
    if let applicationVersion { value["applicationVersion"] = applicationVersion }
    return value
  }
}

/// The synchronous answer to submitting one observation.
///
/// There is deliberately no member meaning validated, verified, confirmed, or
/// entitled. Acceptance means the observation is well formed and queued and
/// asserts nothing about the transaction being real.
public enum MosaicTransactionObservationOutcome: Sendable, Equatable {
  case acceptedForValidation
  case duplicate
  case permanentlyRejected(code: String)
  case retryableFailure(code: String)
}

public struct MosaicTransactionObservationDiagnostics: Sendable, Equatable {
  public let mode: MosaicTransactionObservationMode
  public let queuedCount: Int
  public let acceptedForValidationCount: UInt64
  public let duplicateCount: UInt64
  public let permanentlyRejectedCount: UInt64
  public let retryCount: UInt64
  public let droppedCount: UInt64
  public let lastSafeCode: String?
  public let isFlushInFlight: Bool

  static let disabled = MosaicTransactionObservationDiagnostics(
    mode: .disabled, queuedCount: 0, acceptedForValidationCount: 0, duplicateCount: 0,
    permanentlyRejectedCount: 0, retryCount: 0, droppedCount: 0,
    lastSafeCode: nil, isFlushInFlight: false)
}

public enum MosaicTransactionObservationFlushResult: Sendable, Equatable {
  case delivered(removed: Int, retained: Int)
  case empty
  case deferred
  case disabled
}

/// Receives observations from a commerce provider.
///
/// `enqueue` is synchronous on purpose. A purchase must never suspend on
/// observation handoff, and a `nonisolated func … async` requirement would
/// still allow a provider actor to suspend on a file write.
public protocol MosaicTransactionObservationSink: Sendable {
  func enqueue(_ observation: MosaicTransactionObservation)
}

/// The decoded observation submission result.
struct MosaicTransactionObservationSubmissionResult: Sendable, Equatable {
  let outcome: MosaicTransactionObservationOutcome
  /// Present only on a retryable failure, per the contract.
  let retryAfterSeconds: Int?
}

/// Encodes a submission and decodes the Billing Ingestion Contract 1
/// `observationSubmissionResult` record.
///
/// The decoder is deliberately closed and pessimistic: any status it does not
/// recognize, any envelope it does not recognize, and any response it cannot
/// read become a retryable failure. There is no input that can produce a
/// validated state.
enum MosaicTransactionObservationCodec {
  static let contractVersion = "1"
  static let observationRecordType = "clientTransactionObservation"
  static let recordType = "observationSubmissionResult"

  /// The complete set of keys a submitted record may carry. Adding a key here
  /// is a contract change, not an implementation detail.
  static let envelopeKeys: Set<String> = [
    "billingIngestionContractVersion", "recordType", "payload",
  ]

  /// A client never asserts a Store Environment, a tenant identity, a price,
  /// or an entitlement: the server classifies and resolves those during
  /// validation. `correlation` is present only when a join handle exists.
  static let wireKeys: Set<String> = [
    "observationId", "submissionId", "providerId", "storePlatform", "transactionReference",
    "observedAt", "sourceAuthority", "context", "correlation",
  ]

  static func encode(
    _ observation: MosaicTransactionObservation,
    context: MosaicTransactionObservationContext
  ) throws -> Data {
    var payload: [String: Any] = [
      "observationId": observation.observationID,
      "submissionId": observation.submissionID,
      "providerId": observation.providerID,
      "storePlatform": observation.storePlatform.rawValue,
      "transactionReference": [
        "referenceKind": observation.referenceKind.rawValue,
        "value": observation.reference,
      ],
      "observedAt": MosaicAnalyticsRuntime.timestamp(observation.observedAt),
      // A client observation can only trigger validation, never author a fact.
      "sourceAuthority": "client_observation",
      "context": context.wireValue,
    ]
    if let correlation = observation.correlation {
      payload["correlation"] = correlation.wireValue
    }
    return try JSONSerialization.data(
      withJSONObject: [
        "billingIngestionContractVersion": contractVersion,
        "recordType": observationRecordType,
        "payload": payload,
      ],
      options: [.sortedKeys])
  }

  static func decodeResult(
    _ data: Data, submissionID: String
  ) -> MosaicTransactionObservationSubmissionResult {
    func retryable(_ code: String) -> MosaicTransactionObservationSubmissionResult {
      .init(outcome: .retryableFailure(code: code), retryAfterSeconds: nil)
    }
    guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    else { return retryable("unreadable_submission_result") }
    // A record of another contract version or another type carries different
    // semantics and is never reinterpreted under this one's rules.
    guard root["billingIngestionContractVersion"] as? String == contractVersion,
      root["recordType"] as? String == recordType,
      let payload = root["payload"] as? [String: Any],
      let status = payload["status"] as? String
    else { return retryable("unrecognized_submission_result") }
    guard payload["submissionId"] as? String == submissionID else {
      return retryable("submission_id_mismatch")
    }
    let code = safeCode(payload["code"] as? String)
    switch status {
    case "accepted_for_validation":
      return .init(outcome: .acceptedForValidation, retryAfterSeconds: nil)
    case "duplicate":
      return .init(outcome: .duplicate, retryAfterSeconds: nil)
    case "permanently_rejected":
      return .init(
        outcome: .permanentlyRejected(code: code ?? "permanently_rejected"),
        retryAfterSeconds: nil)
    case "retryable_failure":
      let retryAfter = payload["retryAfterSeconds"] as? Int
      return .init(
        outcome: .retryableFailure(code: code ?? "retryable_failure"),
        retryAfterSeconds: retryAfter.map { min(max($0, 1), 86_400) })
    default:
      return retryable("unrecognized_submission_outcome")
    }
  }

  /// Server codes are echoed into diagnostics, so they are bounded and
  /// stripped of control characters before they are stored.
  private static func safeCode(_ value: String?) -> String? {
    guard let value, !value.isEmpty else { return nil }
    let sanitized = value.unicodeScalars.filter { $0.value >= 0x20 && $0.value != 0x7F }
    guard !sanitized.isEmpty else { return nil }
    return String(String.UnicodeScalarView(sanitized.prefix(96)))
  }
}
